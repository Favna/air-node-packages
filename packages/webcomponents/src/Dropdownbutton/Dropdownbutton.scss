.buttonOverwrites {
  margin-right: 0 !important;
}

.labelOverwrites {
  justify-content: end;
  width: 12rem;
}
