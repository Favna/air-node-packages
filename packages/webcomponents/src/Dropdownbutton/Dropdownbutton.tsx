import ButtonGroup from '@material-ui/core/ButtonGroup';
import ClickAwayListener from '@material-ui/core/ClickAwayListener';
import Grid from '@material-ui/core/Grid';
import Grow from '@material-ui/core/Grow';
import MenuItem from '@material-ui/core/MenuItem';
import MenuList from '@material-ui/core/MenuList';
import Paper from '@material-ui/core/Paper';
import Popper from '@material-ui/core/Popper';
import ArrowDropDownIcon from '@material-ui/icons/ArrowDropDown';
import React, { FC, MouseEvent as ReactMouseEvent, useRef, useState, ReactNode, memo } from 'react';
import { customCss, dataQa, cutText } from '../constants';
import Button from '../Button/Button';
import css from './Dropdownbutton.scss';
import classnames from 'classnames';

export interface DropdownbuttonProps {
  /** The options to show in the dropdown menu */
  options: string[];
  /** The icon displayed in the button to the right. This button is clicked to trigger {@link onClick()} */
  ButtonIcon: ReactNode;
  /** Any options to disabled, using zero-based index for {@link options} */
  disabledOptionIds?: number[];
  /** The color type of the buttons @default primary */
  color?: 'primary' | 'secondary' | 'default' | 'inherit';
  /** The variant of the buttons @default contained */
  variant?: 'text' | 'outlined' | 'contained' | undefined;
  /** Data-qa tags to apply to the button group */
  buttonGroupDataQa?: dataQa;
  /** Data-qa tags to apply to the dropdown button */
  dropdownButtonDataQa?: dataQa;
  /** Data-qa tags to apply to the icon button */
  iconButtonDataQa?: dataQa;
  /** Data-qa tags to apply to each menu item */
  menuItemDataQa?: dataQa;
  /** Data-qa tags to apply to the popper */
  popperDataQa?: dataQa;
  /** Custom CSS classes to pass to the dropdown button */
  dropdownButtonCustomClasses?: customCss;
  /** Custom CSS classes to pass to the icon button */
  iconButtonCustomClasses?: customCss;
  /** Custom CSS classes to pass to the buttonGroup */
  buttonGroupCustomClasses?: customCss;
  /** The action to trigger when the button with {@link ButtonIcon} is pressed. Receives the currently selected option as first parameter */
  onClick(selectedOption: string): unknown;
}

/**
 * Creates a button using pre-defined Rijkswaterstaat styling
 *
 * ```jsx
 * import { Dropdownbutton } from '@rws-air/webcomponents'
 * import { CloudDownload } from '@material-ui/icons';
 *
 * <Dropdownbutton ButtonIcon={<CloudDownload/>} onClick={() => undefined} options={[ 'one', 'two' ]} />
 * ```
 */
export const Dropdownbutton: FC<DropdownbuttonProps> = props => {
  const [ open, setOpen ] = useState(false);
  const anchorRef = useRef<HTMLDivElement>(null);
  const [ selectedIndex, setSelectedIndex ] = useState(0);

  const handleMenuItemClick = (_event: ReactMouseEvent<HTMLLIElement, MouseEvent>, index: number) => {
    setSelectedIndex(index);
    setOpen(false);
  };

  const handleToggle = () => {
    setOpen(prevOpen => !prevOpen);
  };

  const handleClose = (event: ReactMouseEvent<Document, MouseEvent>) => {
    if (anchorRef.current && anchorRef.current.contains(event.target as HTMLElement)) {
      return;
    }
    setOpen(false);
  };

  return (
    <Grid container direction='column' alignItems='center'>
      <Grid item xs={12}>
        <ButtonGroup
          variant={props.variant || 'contained'}
          color={props.color || 'primary'}
          ref={anchorRef}
          aria-label='split button'
          data-qa={props.buttonGroupDataQa}
          className={classnames(props.buttonGroupCustomClasses)}
        >
          <Button
            variant={props.variant || 'contained'}
            color={props.color || 'primary'}
            size='small'
            aria-owns={open ? 'menu-list-grow' : undefined}
            aria-haspopup='true'
            onClick={handleToggle}
            startIcon={<ArrowDropDownIcon />}
            label={cutText(props.options[selectedIndex], 30)}
            className={css.buttonOverwrites}
            customlabelclasses={css.labelOverwrites}
            data-qa={props.dropdownButtonDataQa}
            customclasses={props.dropdownButtonCustomClasses}
          />
          <Button
            variant={props.variant || 'contained'}
            color={props.color || 'primary'}
            onClick={() => props.onClick(props.options[selectedIndex])}
            label={props.ButtonIcon}
            className={css.buttonOverwrites}
            data-qa={props.iconButtonDataQa}
            customclasses={props.iconButtonCustomClasses}
          />
        </ButtonGroup>
        <Popper
          open={open}
          data-qa={props.popperDataQa}
          anchorEl={anchorRef.current}
          transition
          placement='bottom-start'
          disablePortal
        >
          {({ TransitionProps, placement }) => (
            <Grow
              {...TransitionProps}
              style={{
                transformOrigin:
                  placement === 'bottom' ? 'center top' : 'center bottom',
              }}
            >
              <Paper id='menu-list-grow'>
                <ClickAwayListener onClickAway={handleClose}>
                  <MenuList>
                    {props.options.map((option, index) => (
                      <MenuItem
                        data-qa={`${props.menuItemDataQa}-${index}`}
                        key={option}
                        disabled={props.disabledOptionIds ? props.disabledOptionIds.includes(index) : false}
                        selected={index === selectedIndex}
                        onClick={event => handleMenuItemClick(event, index)}
                      >
                        {option}
                      </MenuItem>
                    ))}
                  </MenuList>
                </ClickAwayListener>
              </Paper>
            </Grow>
          )}
        </Popper>
      </Grid>
    </Grid>
  );
};

export default memo(Dropdownbutton);