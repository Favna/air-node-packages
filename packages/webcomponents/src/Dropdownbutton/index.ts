export { default as Dropdownbutton, DropdownbuttonProps } from './Dropdownbutton';