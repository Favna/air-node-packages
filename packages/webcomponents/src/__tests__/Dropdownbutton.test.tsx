import Dropdownbutton from '../Dropdownbutton/Dropdownbutton';
import { shallow, ShallowWrapper } from 'enzyme';
import CloudDownload from '@material-ui/icons/CloudDownload';
import React from 'react';

const mockButtonClick = jest.fn();

describe('Component Tests', () => {
  let dropdownButton: ShallowWrapper;

  beforeAll(() => {
    dropdownButton = shallow(
      <Dropdownbutton
        ButtonIcon={<CloudDownload/>}
        onClick={mockButtonClick}
        options={[ 'one', 'two' ]}
        menuItemDataQa='menu-item'
        iconButtonDataQa='icon-button'
        dropdownButtonDataQa='dropdown-button'
      />
    );
  });

  test('button function called onClick', () => {
    const iconButton = dropdownButton.find('[data-qa="icon-button"]');

    iconButton.simulate('click');

    expect(mockButtonClick).toHaveBeenCalledWith('one');
    expect(mockButtonClick).toHaveBeenCalledTimes(1);
  });
});

describe('Snapshot Testing', () => {
  test('Required Props', () => {
    const dropdownButton = shallow(<Dropdownbutton ButtonIcon={<CloudDownload/>} onClick={mockButtonClick} options={[ 'one', 'two', 'three' ]} />);
    expect(dropdownButton).toMatchSnapshot();
  });

  test('Optional Props', () => {
    const dropdownButton = shallow(
      <Dropdownbutton
        ButtonIcon={<CloudDownload/>}
        onClick={mockButtonClick}
        options={[ 'one', 'two', 'three' ]}
        disabledOptionIds={[ 2 ]}
        color='secondary'
        variant='outlined'
      />
    );
    expect(dropdownButton).toMatchSnapshot();
  });
});