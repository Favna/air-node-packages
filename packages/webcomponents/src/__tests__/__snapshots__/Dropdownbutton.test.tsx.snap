// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing Optional Props 1`] = `
<WithStyles(ForwardRef(Grid))
  alignItems="center"
  container={true}
  direction="column"
>
  <WithStyles(ForwardRef(Grid))
    item={true}
    xs={12}
  >
    <WithStyles(ForwardRef(ButtonGroup))
      aria-label="split button"
      color="secondary"
      variant="outlined"
    >
      <Memo(Button)
        aria-haspopup="true"
        color="secondary"
        label="one"
        onClick={[Function]}
        size="small"
        startIcon={<Memo(ArrowDropDownIcon) />}
        variant="outlined"
      />
      <Memo(Button)
        color="secondary"
        label={<Memo(CloudDownloadIcon) />}
        onClick={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        variant="outlined"
      />
    </WithStyles(ForwardRef(ButtonGroup))>
    <ForwardRef(Popper)
      anchorEl={null}
      disablePortal={true}
      open={false}
      placement="bottom-start"
      transition={true}
    >
      <Component />
    </ForwardRef(Popper)>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;

exports[`Snapshot Testing Required Props 1`] = `
<WithStyles(ForwardRef(Grid))
  alignItems="center"
  container={true}
  direction="column"
>
  <WithStyles(ForwardRef(Grid))
    item={true}
    xs={12}
  >
    <WithStyles(ForwardRef(ButtonGroup))
      aria-label="split button"
      color="primary"
      variant="contained"
    >
      <Memo(Button)
        aria-haspopup="true"
        color="primary"
        label="one"
        onClick={[Function]}
        size="small"
        startIcon={<Memo(ArrowDropDownIcon) />}
        variant="contained"
      />
      <Memo(Button)
        color="primary"
        label={<Memo(CloudDownloadIcon) />}
        onClick={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        variant="contained"
      />
    </WithStyles(ForwardRef(ButtonGroup))>
    <ForwardRef(Popper)
      anchorEl={null}
      disablePortal={true}
      open={false}
      placement="bottom-start"
      transition={true}
    >
      <Component />
    </ForwardRef(Popper)>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
